### PR TITLE
TYP: Subset of "Improved the type stubs in the _libs directory to help with type checking"

### PR DESCRIPTION
**Pull request by twoertwein**
The two big parts missing from #43744 are interval.pyi and offsets.pyi as they result in many mypy errors.

I had already made changes to somewhat accommodate for interval.pyi and offsets.pyi (if these files are included, we have now "only" ~100 mypy errors). I kept these changes in this PR.

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -912,7 +912,11 @@ def maybe_upcast(
     # We get a copy in all cases _except_ (values.dtype == new_dtype and not copy)
     upcast_values = values.astype(new_dtype, copy=copy)
 
-    return upcast_values, fill_value
+    # error: Incompatible return value type (got "Tuple[ndarray[Any, dtype[Any]],
+    # Union[Union[str, int, float, bool] Union[Period, Timestamp, Timedelta, Any]]]",
+    # expected "Tuple[NumpyArrayT, Union[Union[str, int, float, bool], Union[Period,
+    # Timestamp, Timedelta, Any]]]")
+    return upcast_values, fill_value  # type: ignore[return-value]
```
Comment (twoertwein) [inline pandas/core/dtypes/cast.py:919]: Should probably be `np.ndarray` and not be `NumpyArrayT` since the dtype might change.

File: pandas/_libs/tslibs/ccalendar.pyi [hunk 2]
```diff
@@ -8,5 +8,7 @@ def get_firstbday(year: int, month: int) -> int: ...
 def get_lastbday(year: int, month: int) -> int: ...
 def get_day_of_year(year: int, month: int, day: int) -> int: ...
 def get_iso_calendar(year: int, month: int, day: int) -> tuple[int, int, int]: ...
+def is_leapyear(year: int) -> bool: ...
 def get_week_of_year(year: int, month: int, day: int) -> int: ...
 def get_days_in_month(year: int, month: int) -> int: ...
+def dayofweek(y: int, m: int, d: int) -> int: ...
```
Comment (twoertwein) [inline pandas/_libs/tslibs/ccalendar.pyi:14]: `dayofweek` and `is_leapyear` are used only in pyx files

File: pandas/_libs/tslibs/nattype.pyi [hunk 3]
```diff
@@ -13,6 +13,7 @@ iNaT: int
 nat_strings: set[str]
 
 def is_null_datetimelike(val: object, inat_is_null: bool = ...) -> bool: ...
+def checknull_with_nat(val: object) -> bool: ...
```
Comment (twoertwein) [inline pandas/_libs/tslibs/nattype.pyi:16]: Used only in pyx files

File: pandas/_libs/tslibs/np_datetime.pyi [hunk 4]
```diff
@@ -0,0 +1 @@
+class OutOfBoundsDatetime(ValueError): ...
```
Comment (twoertwein) [inline pandas/_libs/tslibs/np_datetime.pyi:1]: again, only in pyx files

File: pandas/_libs/tslibs/period.pyi [hunk 5]
```diff
@@ -14,6 +15,9 @@ from pandas._typing import (
 INVALID_FREQ_ERR_MSG: str
 DIFFERENT_FREQ: str
 
+def is_period_object(obj: object) -> bool: ...
+def get_period_ordinal(dts: datetime, freq: int) -> int: ...
```
Comment (twoertwein) [inline pandas/_libs/tslibs/period.pyi:19]: both are used only in pyx files

File: pandas/tests/tseries/offsets/test_business_day.py [hunk 6]
```diff
@@ -30,7 +30,7 @@
 
 
 class TestBusinessDay(Base):
-    _offset = BDay
+    _offset: type[BDay] = BDay
```
Comment (twoertwein) [inline pandas/tests/tseries/offsets/test_business_day.py:33]: This and some other sub-classes of `Base` require more concrete sub-classes for `_offset`.

This will trigger mypy errors once offset.pyi is added.

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok to me. can you rebase @twoertwein 

@simonjayhawkins @jbrockmendel if comments

**Comment by jreback** [comment]
cc @Dr-Irv

File: pandas/core/dtypes/missing.py [hunk 7]
```diff
@@ -1,8 +1,11 @@
 """
 missing types & inference
 """
+from __future__ import annotations
```
Comment (Dr-Irv) [inline pandas/core/dtypes/missing.py:4]: Is this really needed?  What piece of code is requiring this?

Comment (twoertwein) [reply pandas/core/dtypes/missing.py:4]: Needed for `result: np.ndarray | ExtensionArraySupportsAnyAll`

File: pandas/core/ops/mask_ops.py [hunk 8]
```diff
@@ -46,7 +46,9 @@ def kleene_or(
 
     raise_for_nan(right, method="or")
 
-    if right is libmissing.NA:
+    # error: Non-overlapping identity check (left operand type:
+    # "Union[bool, ndarray[Any, Any]]", right operand type: "NAType")
+    if right is libmissing.NA:  # type: ignore[comparison-overlap]
```
Comment (Dr-Irv) [inline pandas/core/ops/mask_ops.py:51]: If `right` can be `NA`, shouldn't that be reflected in the type signature for this function (`kleene_or`) ?

File: pandas/tests/tseries/offsets/test_business_day.py [hunk 9]
```diff
@@ -1,6 +1,8 @@
 """
 Tests for offsets.BDay
 """
+from __future__ import annotations
```
Comment (Dr-Irv) [inline pandas/tests/tseries/offsets/test_business_day.py:4]: Not clear why this is needed?

**Review by Dr-Irv — COMMENTED** [review]
I see that you put some typing in the `pandas/tests/tseries/...` .   Have we started typing the test code, and running `mypy` on parts of the test code?  

If not, I don't think that should be included in the PR.

**Comment by twoertwein** [comment]
> I see that you put some typing in the `pandas/tests/tseries/...` . Have we started typing the test code, and running `mypy` on parts of the test code?

Some tests are typed checked, including those files.

> If not, I don't think that should be included in the PR.

The test changes I made are technically not needed for this PR. They will be needed for `offsets.pyi` from #43744. Happy to remove them and then add them with `offsets.pyi` later.

File: pandas/tests/tseries/offsets/test_business_day.py (hunk 9 above)
Comment (twoertwein) [reply pandas/tests/tseries/offsets/test_business_day.py:4]: for `type[BDay]`, would need to use `typing.Type[BDay]` otherwise

File: pandas/core/ops/mask_ops.py (hunk 8 above)
Comment (twoertwein) [reply pandas/core/ops/mask_ops.py:51]: Yes, I will look into it.

**Comment by Dr-Irv** [comment]
> The test changes I made are technically not needed for this PR. They will be needed for `offsets.pyi` from #43744. Happy to remove them and then add them with `offsets.pyi` later.

IMHO, I think the changes to the test files should be left out for now.  That will make the PR smaller.  But I would defer to @simonjayhawkins on this

File: pandas/core/ops/mask_ops.py (hunk 8 above)
Comment (twoertwein) [reply pandas/core/ops/mask_ops.py]: `NAType` was probably just forgotten in the type annotations for `kleene_or/xor`: their documentation mentions NA and `kleene_and` allows NA (type annotation and documentation).

File: pandas/core/ops/mask_ops.py [hunk 10]
```diff
@@ -103,23 +99,21 @@ def kleene_xor(
     result, mask: ndarray[bool]
         The result of the logical xor, and the new mask.
     """
+    # To reduce the number of cases, we ensure that `left` & `left_mask`
+    # always come from an array, not a scalar. This is safe, since
+    # A ^ B == B ^ A
     if left_mask is None:
         return kleene_xor(right, left, right_mask, left_mask)
 
+    assert isinstance(left, np.ndarray)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/ops/mask_ops.py:108]: holds in all of our current usages; to be future-safe would need to change this to a check that raises TypeError

[the remark this answers is not in the document]
Comment (twoertwein) [reply pandas/core/ops/mask_ops.py:112]: it raises now a `TypeError`

File: pandas/tests/tseries/offsets/test_business_day.py (hunk 9 above)
Comment (twoertwein) [reply pandas/tests/tseries/offsets/test_business_day.py]: I removed the `pandas/tests/tseries/offsets/test_*.py` changes

**Comment by jreback** [comment]
looks good, can you rebase

@simonjayhawkins if any comments.

**Review by jreback — APPROVED** [review]

File: pandas/_libs/tslibs/ccalendar.pyi (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/_libs/tslibs/ccalendar.pyi:14]: no need to add. Python code cannot call these functions? With these added, Pylance will add to auto-completion. I don't think that's desirable.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @twoertwein

File: pandas/core/ops/mask_ops.py [hunk 11]
```diff
@@ -37,11 +37,13 @@ def kleene_or(
         The result of the logical or, and the new mask.
     """
     # To reduce the number of cases, we ensure that `left` & `left_mask`
-    # always come from an array, not a scalar. This is safe, since because
+    # always come from an array, not a scalar. This is safe, since
     # A | B == B | A
     if left_mask is None:
         return kleene_or(right, left, right_mask, left_mask)
 
+    if not isinstance(left, np.ndarray):
```
Comment (simonjayhawkins) [inline pandas/core/ops/mask_ops.py:45]: can remove the following assert now? (and elsewhere)

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/cast.py:919]: The dtype can change but the Python type (i.e. np.ndarray or subclass) maybe preserved. I'm not sure but the numpy types don't annotate as preserving Python object type so maybe not. This was added for `ma.MaskedArray` handling in the constructors, so could well be that the numpy types are not precise enough, creating this false positive.

File: pandas/core/common.py [hunk 12]
```diff
@@ -515,7 +516,7 @@ def f(x):
 
 
 def convert_to_list_like(
-    values: Scalar | Iterable | AnyArrayLike,
+    values: Scalar | IndexLabel | AnyArrayLike,
```
Comment (simonjayhawkins) [inline pandas/core/common.py:519]: Why is this changed? In code we check for an `Iterable`. `IndexLabel` only includes `Sequence`

File: pandas/core/dtypes/missing.py [hunk 13]
```diff
@@ -51,6 +54,9 @@
 )
 from pandas.core.dtypes.inference import is_list_like
 
+if TYPE_CHECKING:
+    from pandas.core.arrays.base import ExtensionArraySupportsAnyAll
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/missing.py:58]: ExtensionArraySupportsAnyAll is a terrible hack to avoid Protocol. (at a time when we supported Python 3.7)

Probably best to avoid using this for now as the more we use it the greater chance we will run into issues with false positives)

File: pandas/tests/arrays/boolean/test_logical.py [hunk 14]
```diff
@@ -225,3 +230,11 @@ def test_no_masked_assumptions(self, other, all_logical_operators):
             result = getattr(a, all_logical_operators)(other)
             expected = getattr(b, all_logical_operators)(other)
             tm.assert_extension_array_equal(result, expected)
+
+
+@pytest.mark.parametrize("operation", [kleene_or, kleene_xor, kleene_and])
+def test_error_both_scalar(operation):
+    msg = "Either `left` or `right` need to be a np.ndarray."
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/boolean/test_logical.py:237]: should escape the period. (or re.escape the string)

File: pandas/_libs/tslibs/np_datetime.pyi (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/_libs/tslibs/np_datetime.pyi:1]: This is imported in pandas/_libs/tslib.pyx and imported in the main code from there. Should probably update pandas/_libs/tslib.pyi (either import from here and allow re-export or just define there)

**Comment by simonjayhawkins** [comment]
> IMHO, I think the changes to the test files should be left out for now. That will make the PR smaller. But I would defer to @simonjayhawkins on this

The test modules are type checked but the test functions are not (check_untyped_defs is False). Saying that we have several test modules ignored in the mypy config

File: pandas/_libs/tslibs/ccalendar.pyi (hunk 2 above)
Comment (twoertwein) [reply pandas/_libs/tslibs/ccalendar.pyi]: removed

File: pandas/_libs/tslibs/nattype.pyi (hunk 3 above)
Comment (twoertwein) [reply pandas/_libs/tslibs/nattype.pyi]: removed

File: pandas/_libs/tslibs/period.pyi (hunk 5 above)
Comment (twoertwein) [reply pandas/_libs/tslibs/period.pyi]: removed

File: pandas/_libs/missing.pyi [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import numpy as np
+
+from pandas._typing import npt
```
[the remark this answers is not in the document]
Comment (twoertwein) [reply pandas/_libs/missing.pyi]: done

File: pandas/core/ops/mask_ops.py (hunk 11 above)
Comment (twoertwein) [reply pandas/core/ops/mask_ops.py:45]: Removed them, I thought mypy might still need them.

File: pandas/core/common.py (hunk 12 above)
Comment (twoertwein) [reply pandas/core/common.py]: Reverted it. Probably was causing an error with one of the other changes from #43744.

File: pandas/core/dtypes/missing.py (hunk 13 above)
Comment (twoertwein) [reply pandas/core/dtypes/missing.py]: Removed it but added an ignore comment.

File: pandas/tests/arrays/boolean/test_logical.py (hunk 14 above)
Comment (twoertwein) [reply pandas/tests/arrays/boolean/test_logical.py]: escaped

**Comment by jreback** [comment]
@twoertwein can you merge master

**Comment by jreback** [comment]
thanks @twoertwein very nice
